### PR TITLE
Richer Unstructured Plotting Support

### PR DESCRIPTION
**Pull request by bjlittle**
### Description
This PR updates the experimental unstructured plotting support.

Note that, this PR contains highly experimental concept code, in order to demonstrate capability and features from PyVista. It is not tested, but rather "used". Full testing rigor will come later.

This PR touches on quite a few areas, most notably:

* dropping support of py36 (as cirrus-ci is struggling to resolve the py36 dependencies)
* introduces a CLI (via [click](https://click.palletsprojects.com/en/7.x/)) for plotting from the command line entry points i.e.,
   * `iris-pyvista plot ...` to render an unstructured cube from file
   * `iris-pyvista show ...` to render a VTK file
   * `iris-pyvista summary ...` to show a cube summary from file
   * see `iris-pyvista --help` for further details
   * note that `iris-pyvista plot` is the default sub-command
* fixes the rendering delay when using the slider (for a 2D unstructured cube)
* fixes thresholding, which previously was being performed incorrectly
* adds graticule support for both spherical and planar projections
* fixes the `clim` of the `cmap` to cover the full extent of the data (specifically for a 2D time-series)
* supports texture mapping for both spherical and planar projections
* provides over-plotting on a base mesh

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by bjlittle** [comment]
https://user-images.githubusercontent.com/2051656/116331146-0ccca200-a7c7-11eb-93c1-42620b39bb54.mp4

**Comment by bjlittle** [comment]
ezgif com-gif-maker

**Comment by bjlittle** [comment]
ezgif com-gif-maker(1)

**Comment by bjlittle** [comment]
ezgif com-gif-maker(2)

**Comment by bjlittle** [comment]
ezgif com-gif-maker

**Comment by trexfeathers** [comment]
>dropping support of py36 (as cirrus-ci is struggling to resolve the py36 dependencies)

I'm all for this, especially given Numpy's recent py36 deprecation. But since we know one of Iris' main use cases is currently on py36 environments, are we confident that this will move forward soon? Otherwise I'd consider the lack of py36 support a blocker for merging this feature branch.

**Comment by bjlittle** [comment]
`py36` is really on it's last legs... and it's becoming almost impossible to resolve an environment.

Now is the time IMHO.

@jamesp What say ye?

**Comment by jamesp** [comment]
See #4108 for avoiding the conda resolve in CI?  I'm all for chopping py36 going forwards, but we will need to support it for our legacy environments.  I assume those won't go beyond iris 3.0.x though.

**Comment by bjlittle** [comment]
ezgif com-gif-maker(1)
**Animation 4:** Rendering a 2D unstructured cube time-series on a sphere with a base mesh

The land-mask NaNs are thresholded out along with Sea Surface Temperature data out side a specific range. Also showing a stock image texture map, with labelled parallels and meridians graticule.

As the thresholding applies to different cells for different parts of the time-series, the slider shows the associated interactive mesh animation. Underneath a base mesh is rendered.

**Comment by jamesp** [comment]
Looking at the CI tests here though, py37 is timing out

**Comment by bjlittle** [comment]
@jamesp Let's get your PR for `conda-lock` merged on `master`, then pull it into the feature branch.

This seems like the best way forwards here...

**Comment by bjlittle** [comment]
>  I'm all for chopping py36 going forwards, but we will need to support it for our legacy environments. I assume those won't go beyond iris 3.0.x though.

@jamesp The release feature branches e.g., `2.3.x`, `3.0.x` etc are all self-contained and will continue to test against `py36`, given that it still resolves, so that's not a problem going forwards.

I'm more than happy, and relieved, to be dropping `py36` support for `iris`... it's well overdue

File: requirements/ci/py38.yml [hunk 1]
```diff
@@ -31,6 +34,7 @@ dependencies:
   - nc-time-axis
   - pandas
   - pip
+  - pykdtree
```
Comment (trexfeathers) [inline requirements/ci/py38.yml:37]: This should also be included in [requirements/all.txt](https://github.com/SciTools/iris/blob/e74356961f6d6d944c16503392eefb596bb548b5/requirements/all.txt)

File: lib/iris/__init__.py [hunk 2]
```diff
@@ -114,6 +115,7 @@ def callback(cube, field, filename):
     "load_cube",
     "load_cubes",
     "load_raw",
+    "main",
```
Comment (trexfeathers) [inline lib/iris/__init__.py:118]: `main()` was an OK name for the function within `iris.experimental.ugrid.cli`, but given the wider usage could we come up with something less vague? `cli`, or `cli_main`, or something else?

**Review by trexfeathers — CHANGES_REQUESTED** [review]
OK @bjlittle, here's half my review. I'm yet to review `experimental.ugrid.plot`; I'll get on that next.

### CLI review

Great idea! It's really valuable to enable common operations without going through complexities of Python activation, imports etc. - those are big overheads if all you want is a quick summary or visualisation.

But the proposal here offers way **too much configurability** - once a user is invested enough to be tweaking configurations, doing so via Python becomes a much smaller relative overhead.  So there's less benefit to offering the extra config, but we're still paying some classic future maintenance costs, which aren't worth it:

* Extra code complexity when re-visiting.
* Work to keep up to date with the Python functions/classes being used.
* Minor codebase bloat.

I'd prefer that we **keep most of the config in a single place - Python** - rather than maintaining a second entry point for so many options. I've marked the options that I therefore think should be removed (and I've skipped testing that those parts of the code work). Sorry for your sunk cost 😔

I can confirm that the remaining config options work as expected 👍

**Comment by jamesp** [comment]
There is clearly a lot of useful functionality in this CLI, it provides a good overview and easy access to the new plotting code. But maintaining it won't come for free.  

Given the limited user feedback we have so far it would make sense to get this in front of people and see if it is useful.  If it is, great, we can invest proportionally more of our effort in building out this kind of interface.  If it's not used, let's pull it out again and take that as positive feedback to focus our efforts elsewhere.

I suggest we make a condition of a future release (tracked either as a github issue/discussion or Jira) that the utility of the CLI has been reviewed.  Success criteria could be e.g. we have at least one user who is prepared to take the time to write a testimonial entreating us to retain the CLI.  If we can't muster that, then it gets the 🥩.

**Comment by bjlittle** [comment]
This PR may have caused some confusion, so apologies for that.

Just to be clear - not one line of this PR (mostly `cli.py` and `plot.py`) is going to survive the merge from this feature branch to `master`.

It's simply here to bank as part of the internal `3.1a0` engineering tag, for the sole purpose to garner user feedback - an opportunity that we very rarely are afforded.

Aside, from that, this PR has helped me gain a better technical understanding of `pyvista`. Ultimately, this will all grease the wheels for `iris-pyvista`, where all support for unstructured cube mesh rendering will live, and this will be a true opt-in for `iris`, as not all users want or need this capability.

I've written the code with this all in mind, with the functionality being mainly self-contained within `iris.experimental.ugrid.cli` and `iris.experimental.ugrid.plot`, with a wee bit of necessary, but limited connective tissue for requirements and the Python entry-points.

So with regards to `iris`, the life of the code from this PR is most definately ephemeral.

File: lib/iris/experimental/ugrid/plot.py [hunk 3]
```diff
@@ -20,19 +21,46 @@
...
+    "add_graticule_longitude",
     "get_coastlines",
     "plot",
+    "rcParams",
+    "to_ll",
     "to_vtk_mesh",
     "to_xyz",
 ]
 
+# default graticule latitude parallel linspace num
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid/plot.py:43]:
```suggestion
# TODO: a thorough review of this module's code. Not enough time at time of writing - 2021-04-28.

# default graticule latitude parallel linspace num
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]
I'm ceasing my review due to time constraints. There may still be undiscovered gotchas in there somewhere.

Four things that absolutely need addressing before merging:
* https://github.com/SciTools/iris/pull/4109#discussion_r622002395
* https://github.com/SciTools/iris/pull/4109#discussion_r622166991
* https://github.com/SciTools/iris/pull/4109#discussion_r622322726
* https://github.com/SciTools/iris/pull/4109#discussion_r622381309

**Pull request closed without merging**